primitives/float.fil

```diff
@@ -0,0 +1,228 @@
+/**
+ * Defines combinational floating point math operations
+ */
+import "./comb.fil";
+import "./signed.fil";
+
+/// Extracts the sign bit, exponent bits, and mantissa from a floating point number.
+/// Asserts that #WIDTH = #EXPONENT+#MANTISSA+1
+comp _FDecompose[#WIDTH, #EXPONENT, #MANTISSA]<G: 1, ?L: 1=G+1> (
+    @[G, G+1] in: #WIDTH
+) -> (
+    @[G, G+1] sign: 1,
+    @[G, G+1] exponent: #EXPONENT,
+    @[G, G+1] mantissa: #MANTISSA
+) where
+    #WIDTH == #EXPONENT + #MANTISSA + 1,
+    #EXPONENT > 1,
+    #MANTISSA > 1
+{
+    sign_slice := new Select[#WIDTH, #WIDTH-1]<G, G+1>(in);
+    sign = sign_slice.out;
+    exponent_slice := new Slice[#WIDTH, #WIDTH-2, #MANTISSA, #EXPONENT]<G, G+1>(in);
+    exponent = exponent_slice.out;
+    mantissa_slice := new Slice[#WIDTH, #MANTISSA-1, 0, #MANTISSA]<G, G+1>(in);
+    mantissa = mantissa_slice.out;
+}
+
+/// Combines components into a single float
+comp _FCompose[#WIDTH, #EXPONENT, #MANTISSA]<G: 1, ?L: 1=G+1> (
+    @[G, G+1] sign: 1,
+    @[G, G+1] exponent: #EXPONENT,
+    @[G, G+1] mantissa: #MANTISSA
+) -> (
+    @[G, G+1] out: #WIDTH
+) where
+    #WIDTH == #EXPONENT + #MANTISSA + 1,
+    #EXPONENT > 1,
+    #MANTISSA > 1
+{
+    first := new Concat[1, #EXPONENT, 1+#EXPONENT]<G, G+1>(sign, exponent);
+    second := new Concat[1+#EXPONENT, #MANTISSA, #WIDTH]<G, G+1>(first.out, mantissa);
+    out = second.out;
+}
+
+/// Adds two floating point numbers
+comp FAdd[#WIDTH, #EXPONENT, #MANTISSA]<G: 1, ?L: 1=G+1> (
+    @[G, G+1] left: #WIDTH,
+    @[G, G+1] right: #WIDTH,
+) -> (
+    @[G, G+1] out: #WIDTH
+) where
+    #WIDTH == #EXPONENT + #MANTISSA + 1,
+    #EXPONENT > 1,
+    #MANTISSA > 1
+{
+     //// comb stage 1 ////
+    left := new _FDecompose[#WIDTH, #EXPONENT, #MANTISSA]<G, G+1>(left);
+    right := new _FDecompose[#WIDTH, #EXPONENT, #MANTISSA]<G, G+1>(right);
+
+    _one := new Const[1, 1]<G, G+1>();
```

Comment:
Why start these names with an underscore?

Reply:
Components underscored because of a python convention to signify private names, done with actual invokes and instantiations just to help me figure out what exactly should correspond to the verilog.